### ISSUE TITLE
Training and learning rate finder utilities

### ISSUE THREAD

**Issue by SebastienEske** [pytorch/pytorch#25986]
## 🚀 Feature
A function 
```
train_model(model, optimizer, scheduler, dataset, batch_size, shuffle=False, \
                snaphot_prefix = None, num_iters=None, \
                num_epochs=None, iter_size=1, display_iter = 20, snaphost_interval=None, \
                load_snapshot_path=None, restore_dataloader=True, display_gpu=False, schedule_on_iter=False)
```
that allows the user to easily train a model with a dataset without the hassle of coding himself the training loop. It contains the possibility to spread the batch size over several iterations `iter_size`, the possibility to schedule per epoch or per iteration `schedule_on_iter`, the possibility ot train for a set number of iterations or epochs and the possibility to save a reload snapshots containing the model, scheduler, and optimizer states including a few other parameters to properly restart from the snapshot (except for the dataloader since it does not have a state dict).
There is a requirement that the models contains its own loss function.
It does only training, not validation.

Something similar for a learning rate finder.
Both also available for multi node, multi gpu training with automated support for sync batch norm (with simple distributed configuration). It is using DistributedDataParallel with one process per GPU.

And a utility to easily configure learning rates for each parameter: `make_param_groups`.

## Motivation

Some of these features were easily available in caffe and we needed all of them for our work at Pixelz (the company I work for). We have it all implemented already (without unit tests though). We thought it could be useful to the community :-)

We didn't do the validation because we use the snapshots to do it outside of the training instances.

## Pitch
Is anyone interested in this?
Is the scope ok for a single pull request? Or should I make several issues and pull requests?
Can someone help to make the unit tests? (we don't need them ourselves and don't have tons of time available for opensourcing this code)
Does anyone want to add the validation to it?
In which module/submodule should it go?

## Additional context
Example usage:
Configure optimizer and learning rates
```
 custom_lr = { \
        'conv12.weight' : 1e-3, \
        'conv12.bias' : 1e-3  \
    }
    optimizer = optim.SGD(make_param_groups(model, custom_lr), lr=1e-2)
    scheduler = lr_scheduler.StepLR(optimizer, step_size=10, gamma=0.1)
   ```
configure multi gpu settings on a single node and train
```
    os.environ["RANK"] = "0"
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = "35003"
    os.environ["WORLD_SIZE"] = "1"
    
    model = train_model_multigpu(model, optimizer, scheduler, train_dataset, \
batch_size=6, shuffle=False, num_epochs=5, iter_size=2, schedule_on_iter=False, \
snaphost_interval=0.5, snaphot_prefix='/data/pixelz_train_6/snapshot')
```
So far we used it for single or multi gpu on a single node, with several computer vision architectures (VGG, resnet and variants, etc) without problem.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @vincentqb

**Comment by zou3519** [comment]
I'm not sure that this one single API (`train_model` and `train_model_multigpu` can work for all use cases). I remember there were discussions in the past about not simplifying the training loop for users because there was no one-size-fits-all solution (cc @soumith) but that was a while back and maybe we can revisit the question.

**Comment by SebastienEske** [comment]
Well, it obviously can't fit all cases because there are so many of them. Maybe highlighting the limits of the API could help:

- the dataset needs to be of constant type for `torch.utils.data.distributed.DistributedSampler`
- since in the distributed case, we need a specific sampler (and we wanted to handle that automatically), and we wanted to keep the same function signature, the user cannot provide his/her own dataloader and sampler. We could enable the user to provide the dataloader instead of `dataset, batch_size, shuffle=False` either for only `train_model` or for both APIs. In the multi-gpu case, then it would be the user's responsibility to provide the appropriate sampler and he would need to change it between single and multi gpu cases.
- we use the following trick to pass an arbitrary dataset to the neural network so there should not be any other limitation to the dataset
```
for inputs in data_loader:
    loss = model(*inputs) / iter_size
```
- the model forward function should only take as inputs the object(s) return by a dataset iteration in the same order. If the model needs other variables that are constant throughout the training, they can be passed at model initialization. If the other variables depend on factors unrelated to the model, they (or the factors) can be in the dataset and if they vary based on the model, then the model can compute them internally.
- we only tested the case where the model returns a single one valued tensor (e.g. a single number, the loss) but when going over the code, there should not be a problem as long as the model returns a single tensor (not necessarily single valued) since this value is only used to compute and display the loss.
- the API only supports a single optimizer (cannot use two optimizers at the same time) but I don't see this as a strong limitation.

If you already have an idea of where I should put the code (if this feature is accepted), let me know and I will upload it on my branch/fork of the repo in the right place so you can see the code for yourself. Maybe in `torch.utils.training` (new submodule)?

**Comment by fmassa** [comment]
Hi,

Thanks for opening the issue and raising this discussion!

We used to have a trainer in PyTorch (with plugin abstractions etc) that we never actually advertised nor recommended using. It wasn't generic enough (we couldn't write all the examples in `pytorch/examples` using it!), and we decided to remove it in https://github.com/pytorch/pytorch/pull/12487

From my point of view, trainers should live together with application code (or domain-specific frameworks). For example, an object detection framework should have its own trainer, an image classification framework might have a different need wrt the trainer, etc.

Plus, a very common point for all training loops is logging, and we do not have a blessed way of doing logging in PyTorch - it's up to the user to setup what they want, be it `print` / `logging` / `tensorboard` / something else. 
[`ignite`](https://github.com/pytorch/ignite), which provides abstractions to write the training loop and more, and lives outside of PyTorch core.

For those reasons, I believe we would not want to have a trainer abstraction in core PyTorch.

Let me know if you disagree.

**Comment by SebastienEske** [comment]
Hello @fmassa I completely agree that if it is not generic enough then a trainer will not be useful. At the same time some other frameworks have it without problem (caffe, tensorflow...). And it is a common criticism of pytorch that users need to code a lot of boilerplate code before actually getting something out of it. Another concern would be the opportunity for bugs if everyone codes its own training and evaluation loop (of course they can reuse the numerous examples).

I also agree that some code is highly application specific and that's partially why we haven't done the validation.

I believe the API I propose to be quite simple and flexible but again, you are right that it should be generic enough. Being able to run most if not all the examples with it sounds like a good criteria to me. I'll have a look at them and see if this is actually possible. If not, then, yes let's keep this closed. If it is possible, would you reconsider?

**Comment by SebastienEske** [comment]
Hello again,

I went through the examples and as you probably expected found a few limitations with the proposed API. It cannot do the following:
- criterion to stop when loss is too low (easy to add as yet another param)
- optimizer closure (could be added as a parameter? can we define the closure outside of the training loop?)
- gradient clipping (not trivial to add as the generic functionality is adding the possibility to have some extra processing function at the end of each step/epoch)
- dataset reset for torchtext after each epoch (same issue as grad clipping)
- hogwild and other custom distributed settings (not even trying to add this)
- gan (that's a case on its own, not trying it)
- reinforcement learning would be possible but ugly because of the need to include [part of the training loop](https://github.com/pytorch/examples/blob/ee964a2eeb41e1712fe719b83645c79bcbd0ba1a/reinforcement_learning/reinforce.py#L85) in the network

So yes, that's quite a few unhandled cases... 2 are easy to fix, 2 are much harder (actually the same one) and 2 will not be fixed, out of 12 examples. 

Probably not good enough...

**Comment by SebastienEske** [comment]
One more question though: do you want to salvage the `make_param_groups` utility? It allows you to only pass the parameters for which you want a custom learning rate and will return the proper groups for the optimizer. It will also put the parameters with the same learning rates in the same group. So in the example below there will be two groups.
```
custom_lr = { \
        'conv12.weight' : 1e-3, \
        'conv12.bias' : 1e-3  \
}
optimizer = optim.SGD(make_param_groups(model, custom_lr), lr=1e-2)
```

**Comment by soumith** [comment]
the examples make a good start, and there's a lot more that the examples dont cover.

Promising trainer abstractions in the pytorch community are:

- ignite: http://github.com/pytorch/ignite
- lightning: https://github.com/williamFalcon/pytorch-lightning
- skorch: https://github.com/skorch-dev/skorch

There are a few nice ones, but didn't get as much traction.

**Comment by SebastienEske** [comment]
Ok, thanks for the pointers!
